### PR TITLE
Implementation of EOF in Solidity [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by rodiazet**
[EOF Mega Spec](https://github.com/ipsilon/eof/blob/main/spec/eof.md)

- [EIP-7480 ](https://eips.ethereum.org/EIPS/eip-7480)(Only `DATALOADN`)
- [EIP-4200](https://eips.ethereum.org/EIPS/eip-4200) (w/o RJUMPV)
- [EIP-4750](https://eips.ethereum.org/EIPS/eip-4750)
- [EIP-6206](https://eips.ethereum.org/EIPS/eip-6206)
- [EIP-7620](https://eips.ethereum.org/EIPS/eip-7620)
- [EIP-7069](https://eips.ethereum.org/EIPS/eip-7069)
- [EIP-5450](https://eips.ethereum.org/EIPS/eip-5450)
- evmc update to 12
- Change `sendMessage` interface
- Add `compileToEOF` flag in semantic tests settings
- Semantic tests updated

TODO: 
- There is a couple TODOs added in the commits. They need to be implemented in separated PR.
- ~~Immutables bug reported by Paradigm needs to be fixed.~~ FIXED
- Semantic tests may stop passing because part of them depend on exact contract code which may change with compiler development. Additional support for contract addresses needs to be implemented in `soltest`.
- CI integration may be required.

### CODE REVIEW

File: libevmasm/Assembly.cpp [hunk 1]
```diff
@@ -58,35 +59,37 @@ AssemblyItem const& Assembly::append(AssemblyItem _i)
 {
 	assertThrow(m_deposit >= 0, AssemblyException, "Stack underflow.");
 	m_deposit += static_cast<int>(_i.deposit());
-	m_items.emplace_back(std::move(_i));
-	if (!m_items.back().location().isValid() && m_currentSourceLocation.isValid())
-		m_items.back().setLocation(m_currentSourceLocation);
-	m_items.back().m_modifierDepth = m_currentModifierDepth;
-	return m_items.back();
+	auto& currentItems = m_codeSections.at(m_currentCodeSection).items;
```
Comment (nikola-matic) [inline libevmasm/Assembly.cpp:63]: `m_codeSection.at(...)` can throw - either a `util::contains` check should go here, or even better, an assertion if you're sure that the current code section is present.

File: libevmasm/Assembly.cpp [hunk 2]
```diff
@@ -58,35 +59,37 @@ AssemblyItem const& Assembly::append(AssemblyItem _i)
...
+	currentItems.back().m_modifierDepth = m_currentModifierDepth;
+	return currentItems.back();
 }
 
 unsigned Assembly::codeSize(unsigned subTagSize) const
 {
 	for (unsigned tagSize = subTagSize; true; ++tagSize)
 	{
 		size_t ret = 1;
-		for (auto const& i: m_data)
```
Comment (nikola-matic) [inline libevmasm/Assembly.cpp:73]: Is `m_data` no longer taken into account?

File: libevmasm/Assembly.cpp [hunk 3]
```diff
@@ -700,16 +711,21 @@ std::map<u256, u256> const& Assembly::optimiseInternal(
...
 
 	// Run optimisation for sub-assemblies.
+	// TODO: verify and double-check this for EOF.
 	for (size_t subId = 0; subId < m_subs.size(); ++subId)
 	{
 		OptimiserSettings settings = _settings;
 		Assembly& sub = *m_subs[subId];
+		std::set<size_t> referencedTags;
+		for (auto& codeSection: m_codeSections)
+			referencedTags += JumpdestRemover::referencedTags(codeSection.items, subId);
```
Comment (nikola-matic) [inline libevmasm/Assembly.cpp:744]: @ekpyron didn't you say that we are no longer going to have JUMPDESTS in EOF? Relative static jumps and all that jazz?

File: libevmasm/Assembly.cpp [hunk 4]
```diff
@@ -834,6 +862,63 @@ std::map<u256, u256> const& Assembly::optimiseInternal(
...
+		size_t i = worklist.top();
+		worklist.pop();
+		AssemblyItem const& item = _items.at(i);
+		size_t stack_height_change = item.deposit();
+		ptrdiff_t stackHeight = stack_heights.at(i);
+		assertThrow(stackHeight != -1, AssemblyException, "");
+
+		std::vector<size_t> successors;
+
+		if (
```
Comment (nikola-matic) [inline libevmasm/Assembly.cpp:912]: I assume that `ConditionalRelativeJump` is not here because both 'branches' need to have their stack heights calculated? I.e. both the next assembly item as well as the tagged one that we're (possibly) jumping to?

File: libevmasm/Assembly.cpp [hunk 5]
```diff
@@ -834,6 +862,63 @@ std::map<u256, u256> const& Assembly::optimiseInternal(
 	return *m_tagReplacements;
 }
 
+namespace
+{
+uint16_t calcMaxStackHeight(std::vector<AssemblyItem> const& _items, uint16_t _args)
+{
+	uint16_t maxStackHeight = _args;
+	std::stack<size_t> worklist;
+	std::vector<int32_t> stack_heights(_items.size(), -1);
```
Comment (nikola-matic) [inline libevmasm/Assembly.cpp]:
```suggestion
	std::vector<size_t> stack_heights(_items.size(), std::numeric_limits<size_t>::max);
```

`stackHeight` should also be unsigned - I would assume our stack height is never going to be negative?

File: libevmasm/Assembly.cpp [hunk 6]
```diff
@@ -834,6 +862,63 @@ std::map<u256, u256> const& Assembly::optimiseInternal(
 	return *m_tagReplacements;
 }
 
+namespace
+{
+uint16_t calcMaxStackHeight(std::vector<AssemblyItem> const& _items, uint16_t _args)
```
Comment (nikola-matic) [inline libevmasm/Assembly.cpp]:
```suggestion
uint16_t calculateMaxStackHeight(std::vector<AssemblyItem> const& _items, uint16_t _args)
```

This isn't rust, we don't have to use abbreviations.

File: libevmasm/Assembly.cpp [hunk 7]
```diff
@@ -834,6 +862,63 @@ std::map<u256, u256> const& Assembly::optimiseInternal(
...
+			auto it = std::find(_items.begin(), _items.end(), item.tag());
+			assertThrow(it != _items.end(), AssemblyException, "Tag not found.");
+			successors.emplace_back(static_cast<size_t>(std::distance(_items.begin(), it)));
+		}
+
+		maxStackHeight = std::max(maxStackHeight, static_cast<uint16_t>(stackHeight + static_cast<ptrdiff_t>(item.maxStackHeightDelta())));
+		stackHeight += static_cast<ptrdiff_t>(stack_height_change);
+
+		for (size_t s: successors)
+		{
```
Comment (nikola-matic) [inline libevmasm/Assembly.cpp:936]: I would add an assert here that `s < stack_heights.size()`.

File: libevmasm/Assembly.cpp [hunk 8]
```diff
@@ -867,175 +954,364 @@ LinkerObject const& Assembly::assemble() const
...
+				pushesImmutables = true;
 	if (setsImmutables || pushesImmutables)
 		assertThrow(
 			setsImmutables != pushesImmutables,
 			AssemblyException,
 			"Cannot push and assign immutables in the same assembly subroutine."
 		);
 
+	assertThrow(!m_codeSections.empty(), AssemblyException, "Expected at least one code section.");
+	assertThrow(eof || m_codeSections.size() == 1, AssemblyException, "Expected exactly one code section in non-EOF code.");
```
Comment (nikola-matic) [inline libevmasm/Assembly.cpp:1003]:
```suggestion
	assertThrow(!eof && m_codeSections.size() == 1, AssemblyException, "Expected exactly one code section in non-EOF code.");
```

Not a big deal, but I feel this is more explicit.

File: libevmasm/BlockDeduplicator.cpp [hunk 9]
```diff
@@ -131,7 +131,7 @@ BlockDeduplicator::BlockIterator& BlockDeduplicator::BlockIterator::operator++()
 {
 	if (it == end)
 		return *this;
-	if (SemanticInformation::altersControlFlow(*it) && *it != AssemblyItem{Instruction::JUMPI})
+	if (SemanticInformation::altersControlFlow(*it) && *it != AssemblyItem{Instruction::JUMPI} && it->type() != ConditionalRelativeJump)
```
Comment (nikola-matic) [inline libevmasm/BlockDeduplicator.cpp:134]:
```suggestion
	if (SemanticInformation::altersControlFlow(*it) && *it != AssemblyItem{Instruction::JUMPI} && *it != AssemblyItem{Instruction::RJUMPI})
```

Is this also an option?

File: libevmasm/Assembly.cpp (hunk 3 above)
Comment (ekpyron) [reply libevmasm/Assembly.cpp:744]: We still need tags, though - ``JumpdestRemover`` is maybe not the best name anymore, though - it's really about assembly tags. I'd just spare ourselves a renaming in this PR, though, we can rename the thing later.

File: libevmasm/Assembly.cpp (hunk 8 above)
Comment (ekpyron) [reply libevmasm/Assembly.cpp:1003]: That's not the same and your version will (obviously) fail for non-eof code :-). This function is shared between legacy and eof - the assertion is to make sure that legacy code only has one code section (i.e. either it's eof code, then fine - or otherwise, i.e. it's legacy code, it has to be only one code section).

Comment (rodiazet) [reply libevmasm/Assembly.cpp:1003]: BTW. I would refactor this function and separated common and uncommon fragments for EOF and legacy, but it can be done in additional PR. I can do it. I think I have now good understanding what's happening there :)

File: libevmasm/Assembly.cpp (hunk 3 above)
Comment (ekpyron) [reply libevmasm/Assembly.cpp:744]: Also keep in mind that most of this code is still shared between EOF and legacy.

File: libevmasm/Assembly.cpp [hunk 10]
```diff
@@ -396,7 +399,8 @@ void Assembly::assemblyStream(
 {
 	Functionalizer f(_out, _prefix, _sourceCodes, *this);
 
-	for (auto const& i: m_items)
+	// TODO: support EOF
```
Comment (ekpyron) [inline libevmasm/Assembly.cpp:404]: In cases like this (and e.g. similarly for ``Assembly::assemblyJSON`` and possibly a few of other similar TODOs like this), it'd be better to add an explicit

```c++
solUnimplementedAssert(!m_eofVersion.has_value(), "Assembly output for EOF is not yet implemented.");
```

Better to error out and then implement it properly later in a subsequent PR than to give a partial output of only the first code section with the rest missing.

File: libevmasm/BlockDeduplicator.cpp (hunk 9 above)
Comment (ekpyron) [reply libevmasm/BlockDeduplicator.cpp:134]: Nope :-). Jumps and RJumps are quite different assembly items - JUMP and JUMPI don't carry any additional information, since their target is on stack, so they're - in that sense - not different from any other instruction. And thus can just be created as plain instruction-assembly items. On the other hand, RJUMP* need to carry immediate arguments. We should probably actually *disallow* creating assembly items from RJUMP* instructions (assert against it in the constructor), if that's not done yet - relative jumps without immediate argument are ill-defined, so the comparison here would be equally ill-defined.
The corresponding case for current jumps is the "Tag" that's pushed one assembly item before the actual jump - that's also not a normal instruction and can't be created with any ``Instruction::`` constructor of ``AssemblyItem``.

File: libevmasm/Assembly.cpp (hunk 2 above)
Comment (ekpyron) [reply libevmasm/Assembly.cpp:73]: That part of the calculation moved to ``::assemble`` - it's part of ``bytesRequiredForDataUpperBound`` there (both for EOF and legacy)...

But we should probably rethink this as well at the latest as part of https://github.com/ethereum/solidity/pull/15294#discussion_r1699070143 - For example: since in EOF there won't be any ``PushTag``s, the weird outer loop here becomes superfluous as well, since code size no longer depends on the tag size (the "tag size" in general becomes irrelevant).

For now, maybe a comment in the header is enough, clarifying that even for legacy this now only counts "pure" code size without data?

File: libevmasm/Assembly.cpp (hunk 1 above)
Comment (rodiazet) [reply libevmasm/Assembly.cpp:63]: Right. I'm adding an assertion as `m_currentCodeSection` is changed on in `void beginFunction(uint16_t _functionID)` and similar assertion is there too.

File: libevmasm/Assembly.cpp (hunk 10 above)
Comment (rodiazet) [reply libevmasm/Assembly.cpp:404]: Done. Additionally I switched off the assembly optimization for EOF as event first step requires verification.

File: libevmasm/Assembly.cpp (hunk 4 above)
Comment (rodiazet) [reply libevmasm/Assembly.cpp:912]: Yes. For `ConditionalRelativeJump` we need to calculated both branches and one of them starts with the next instruction but second is the jump target. It's being added to successors by next `if`

File: libevmasm/Assembly.cpp (hunk 5 above)
Comment (rodiazet) [reply libevmasm/Assembly.cpp]: Fixed

File: libevmasm/Assembly.cpp (hunk 7 above)
Comment (rodiazet) [reply libevmasm/Assembly.cpp:936]: Added

File: libevmasm/Assembly.cpp (hunk 6 above)
Comment (rodiazet) [reply libevmasm/Assembly.cpp]: Changed.

File: libevmasm/Assembly.cpp (hunk 8 above)
Comment (rodiazet) [reply libevmasm/Assembly.cpp:1003]: @ekpyron How about moving EOF assembly to separated function (in this PR) and leave the legacy as it is? Would it be easier to review?

File: libsolidity/experimental/codegen/IRGenerator.cpp [hunk 11]
```diff
@@ -72,8 +72,12 @@ std::string IRGenerator::run(
 	Whiskers t(R"(
 		object "<CreationObject>" {
 			code {
-				codecopy(0, dataoffset("<DeployedObject>"), datasize("<DeployedObject>"))
-				return(0, datasize("<DeployedObject>"))
+				<?eof>
+					returncontract("<DeployedObject>", 0, 0)
```
Comment (ekpyron) [inline libsolidity/experimental/codegen/IRGenerator.cpp:76]: Shouldn't auxdata here involve the immutables?

Comment (ekpyron) [reply libsolidity/experimental/codegen/IRGenerator.cpp:76]: Nevermind, I just noticed that this is in ``libsolidity/experimental`` (that's currently largely unmaintained - we'll likely remove it on develop, since experimental solidity development has moved to prototypes in separate repos for now)

File: libyul/backends/evm/ControlFlowGraphBuilder.cpp [hunk 12]
```diff
@@ -76,6 +87,16 @@ void cleanUnreachable(CFG& _cfg)
 		cxx20::erase_if(node->entries, [&](CFG::BasicBlock* entry) -> bool {
 			return !reachabilityCheck.visited.count(entry);
 		});
+
+	// Remove functions which are never referenced.
```
Comment (ekpyron) [inline libyul/backends/evm/ControlFlowGraphBuilder.cpp:91]: This change in ``cleanUnreachable`` could be split out into a separate PR and merged directly into ``develop`` up front, right?
EDIT: we can revisit once we get to the commits containing it, whether splitting really makes things easier, fine for now.

File: libsolidity/interface/StandardCompiler.cpp [hunk 13]
```diff
@@ -837,7 +837,7 @@ std::variant<StandardCompiler::InputsAndSettings, Json> StandardCompiler::parseI
 	{
 		if (!settings["eofVersion"].is_number_unsigned())
 			return formatFatalError(Error::Type::JSONError, "eofVersion must be an unsigned integer.");
-		auto eofVersion = settings["evmVersion"].get<uint8_t>();
+		auto eofVersion = settings["eofVersion"].get<uint8_t>();
```
Comment (ekpyron) [inline libsolidity/interface/StandardCompiler.cpp:840]: This and the related evm-version check change can and should separately go into develop :-).
EDIT: as per chat, let's just merge the first three commits for now :-).

File: libsolidity/interface/CompilerStack.cpp [hunk 14]
```diff
@@ -1412,7 +1420,7 @@ void CompilerStack::compileContract(
 
 	Contract& compiledContract = m_contracts.at(_contract.fullyQualifiedName());
 
-	std::shared_ptr<Compiler> compiler = std::make_shared<Compiler>(m_evmVersion, m_revertStrings, m_optimiserSettings);
+	std::shared_ptr<Compiler> compiler = std::make_shared<Compiler>(m_evmVersion, m_eofVersion, m_revertStrings, m_optimiserSettings);
```
Comment (ekpyron) [inline libsolidity/interface/CompilerStack.cpp:1427]: Instead of passing ``m_eofVersion`` around here, we can just assert
```
solUnimplementedAssert(!m_eofVersion, "EOF is not supported for legacy code generation");
```

(this is only-legacy, see the ``solAssert(!m_viaIR, "");`` below)
And thereby leave ``libsolidity/codegen/Compiler.h`` and ``libsolidity/codegen/CompilerContext.h`` largely untouched (resp. where they need an eof version just use ``std::nullopt`` directly).

File: libevmasm/AssemblyItem.cpp [hunk 15]
```diff
@@ -327,6 +380,30 @@ std::string AssemblyItem::toAssemblyText(Assembly const& _assembly) const
...
+		text = "jumpf(" +  std::to_string(static_cast<size_t>(data())) + ")";
+		break;
+	case RetF:
+		text = "retf";
+		break;
+	case EofCreate:
+		text = "eofcreate(" +  std::to_string(static_cast<size_t>(data())) + ")";
+		break;
+	case ReturnContract:
+		text = "returcontract(" +  std::to_string(static_cast<size_t>(data())) + ")";
```
Comment (cairoeth) [inline libevmasm/AssemblyItem.cpp:402]: typo in assembly text of `RETURNCONTRACT`
```suggestion
		text = "returncontract(" +  std::to_string(static_cast<size_t>(data())) + ")";
```

File: libsolidity/interface/CompilerStack.cpp (hunk 14 above)
Comment (rodiazet) [reply libsolidity/interface/CompilerStack.cpp:1427]: Addressed in https://github.com/ethereum/solidity/pull/15356/files#diff-0f19df92d55a7d7fe38fdf4be1e14b364671c4934853d13a33b31aedf3f917bdR1424

**Comment by rodiazet** [comment]
This PR is splitted into many separated PRs which are in the review. This PRs is supposed to be closed without merging after those PRs are merged.
